ons_alpha/private_media/signal_handlers.py

```diff
@@ -0,0 +1,104 @@
+from django.apps import apps
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Q, QuerySet
+from wagtail.images.models import AbstractImage
+from wagtail.models import PAGE_MODEL_CLASSES, Page, ReferenceIndex
+from wagtail.signals import page_published, page_unpublished
+
+from .models import PrivateMediaCollectionMember
+
+
+def get_private_media_models():
+    return [m for m in apps.get_models() if issubclass(m, PrivateMediaCollectionMember)]
+
+
+def publish_page_media_on_publish(instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to be connected to the 'page_published' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used by the page, and ensuring that it is also made public.
+    """
+    ids_by_ctype = defaultdict(list)
+    for ct_id, obj_id in get_all_page_media_ids(instance):
+        ids_by_ctype[ct_id].append(obj_id)
+
+    for ct_id, id_list in ids_by_ctype.items():
+        model_class = ContentType.objects.get_for_id(ct_id).model_class()
+        id_list = [model_class._meta.pk_field.to_python(id) for id in id_list]
+        private_objects_qs = model_class.objects.filter(id__in=id_list, is_private=True).select_related(
+            "collection", "last_public_collection", "last_private_collection"
+        )
+        if issubclass(model_class, AbstractImage):
+            private_objects_qs = private_objects_qs.prefetch_related("renditions")
+        model_class.objects.bulk_make_public(list(private_objects_qs))
+
+
+def unpublish_page_media_on_unpublish(instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to be connected to the 'page_unpublished' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used exclusively by the page, and ensuring that it is also made
+    private.
+    """
+    ids_by_ctype = defaultdict(list)
+    for ct_id, obj_id in get_unique_to_page_media_ids(instance):
+        ids_by_ctype[ct_id].append(obj_id)
+
+    for ct_id, id_list in ids_by_ctype.items():
+        model_class = ContentType.objects.get_for_id(ct_id).model_class()
+        id_list = [model_class._meta.pk_field.to_python(id) for id in id_list]
+        public_objects_qs = model_class.objects.filter(id__in=id_list, is_private=False).select_related(
+            "collection", "last_public_collection", "last_private_collection"
+        )
+        if issubclass(model_class, AbstractImage):
+            public_objects_qs = public_objects_qs.prefetch_related("renditions")
+        model_class.objects.bulk_make_private(list(public_objects_qs))
+
+
+def register_signal_handlers():
+    for page_model in PAGE_MODEL_CLASSES:
+        page_published.connect(publish_page_media_on_publish, sender=page_model, dispatch_uid="publish_media")
+        page_unpublished.connect(unpublish_page_media_on_unpublish, sender=page_model, dispatch_uid="unpublish_media")
```

Comment:
Suggestion: Can this be done by omitting the `sender` argument instead? Looping over the internal `PAGE_MODEL_CLASSES` feels like it's prone to a race condition, where the signal runs before some classes are loaded.

Reply:
Happy to give that a go!